packages/remirror__extension-entity-reference/src/entity-reference-extension.ts

```diff
@@ -0,0 +1,264 @@
+  createPlugin(): CreateExtensionPlugin {
+    return {
+      state: {
+        init: (_: { [key: string]: any }, state: EditorState) => {
+          const entityReferences = this.getDisjoinedEntityReferences(state.doc);
+          return { entityReferences };
+        },
+        apply: (_tr: Transaction, _value: any, _oldState: EditorState, newState: EditorState) => {
+          const entityReferences = this.getDisjoinedEntityReferences(newState.doc);
+          return { entityReferences };
+        },
+      },
+      props: {
+        decorations: (state: EditorState) => {
+          return createDecorationSet({ state, extension: this });
+        },
+      },
+    };
+  }
+
+  @command()
+  addEntityReference(id?: string): CommandFunction {
+    return ({ state, tr, dispatch }) => {
+      const { from, to } = state.selection;
+      const newId = id ?? (this.options.createId ? this.options.createId() : uniqueId());
+      const newMark = this.type.create({
+        id: newId,
+      });
+      try {
+        tr = tr.addMark(from, to, newMark);
+      } catch (error: any) {
+        throw new Error(
+          `Can't add entityReference ${JSON.stringify(
+            newMark,
+          )} to range {from: ${from}, to: ${to}}: ${error.message}`,
+        );
+      }
+
+      dispatch?.(tr);
+      return true;
+    };
+  }
+
+  @command()
+  removeEntityReference(entityReferenceId: string): CommandFunction {
+    return ({ tr, dispatch }) => {
+      if (!entityReferenceId) {
+        throw new Error(`EntityReference can't be removed without specifying tags its ID!`);
+      }
+
+      const singleMark = this.type.create({ id: entityReferenceId });
+      try {
+        // Remove all entityReference marks with id=entityReferenceId in all the document
+        // Getting the last position https://discuss.prosemirror.net/t/getting-the-last-node-and-its-pos-in-the-document/3091
+        tr = tr.removeMark(1, tr.doc.content.size, singleMark);
+      } catch (error: any) {
+        throw new Error(`Can't remove entityReference ${entityReferenceId}: ${error.message}`);
+      }
+      dispatch?.(tr);
+      return true;
+    };
+  }
+
+  @command()
+  redrawEntityReferences(): CommandFunction {
+    return ({ tr, dispatch }) => {
+      dispatch?.(
+        tr.setMeta(EntityReferenceExtension.name, {
+          type: ActionType.REDRAW_ENTITY_REFERENCES,
+        }),
+      );
+
+      return true;
+    };
+  }
+
+  /**
+   * Get all disjoined entityReference attributes from the document.
+   */
+  @helper()
+  getDisjoinedEntityReferences(doc: ProsemirrorNode): Helper<EntityReferenceMetaData[][]> {
+    const disjoinedEntityReferences: EntityReferenceMetaData[][] = [];
+    doc.descendants((node: Node, pos: number) => {
+      const subDisjoinedEntityReferences = getDisjoinedEntityReferencesFromNode(
+        node,
+        pos,
+        this.name,
+      );
+
+      if (subDisjoinedEntityReferences.length === 0) {
+        return true;
+      }
```

Comment:
I suspect this is an over simplification. I suspect we should 

```ts
if (node.isAtom) {
  return false;
}

if (!node.isText) { // Assuming the only thing that can have marks is text nodes? I might be wrong here
  return true;
}

const subDisjoinedEntityReferences = getDisjoinedEntityReferencesFromNode(
  node,
  pos,
  this.name,
);
```

Reply:
There are some use cases where we might need to reference a non-text node.